storage: preemptive snapshots are being dropped on the floor

See #7332.

During initial replication of the 5 initial ranges, replica reservations are sometimes declined. The error in the logs is:

```
I160621 20:40:59.638047 storage/reservation.go:155  could not book reservation {StoreRequestHeader:{NodeID:3 StoreID:3} FromNodeID:1 FromStoreID:1 RangeID:5 RangeSize:67108864}, not enough available reservation space (requested:67108864, reserved:201326592, maxReserved:5)
```

This differs from my expectation. With one node sending the reservations, I would expect that we'd see a sequence of reserve-replicate-fill. Instead, we somehow saw 5 consecutive reservations with the 5th one failing.

@bdarnell The naive implementation of your suggestion causes the following panic:

```
panic: config.ID must not be zero

goroutine 4555 [running]:
panic(0x5148040, 0xc82077e4d0)
        /Users/pmattis/Development/go-1.6/src/runtime/panic.go:481 +0x3e6
github.com/coreos/etcd/raft.NewRawNode(0xc8207a3980, 0x0, 0x0, 0x0, 0x20, 0x0, 0x0)
        /Users/pmattis/Development/go/src/github.com/coreos/etcd/raft/rawnode.go:74 +0xb4
github.com/cockroachdb/cockroach/storage.(*Replica).withRaftGroupLocked(0xc82030e540, 0xc8207a3b20, 0x0, 0x0)
        /Users/pmattis/Development/go/src/github.com/cockroachdb/cockroach/storage/replica.go:287 +0x1a6
github.com/cockroachdb/cockroach/storage.(*Replica).withRaftGroup(0xc82030e540, 0xc8207a3b20, 0x0, 0x0)
        /Users/pmattis/Development/go/src/github.com/cockroachdb/cockroach/storage/replica.go:325 +0x8e
github.com/cockroachdb/cockroach/storage.(*Store).handleRaftMessage(0xc8200898c0, 0xc820174a20, 0x0, 0x0)
        /Users/pmattis/Development/go/src/github.com/cockroachdb/cockroach/storage/store.go:2061 +0x967
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
        /Users/pmattis/Development/go/src/github.com/cockroachdb/cockroach/storage/store.go:2119 +0x552
...
```

There are two additional problems here. After applying a preemptive snapshot a subsequent raft message will arrive for the replica and we'll go to create the raft group. This creation panics with the following error:

```
panic: [group 1] applied(18) is out of range [prevApplied(10), committed(10)]
```

This panic is caused because committed is initialized from `Replica.FirstIndex` and `Replica.InitialState().Committed`. But snapshots do not contain the raft hard state which is unreplicated, so `Replica.InitialState()` is empty. I have yet to trace through how snapshots work when the raft group has already been created.

The second problem is that the preemptive snapshot request does not wait for the snapshot to be applied remotely. Because we send the snapshot over a gRPC stream, no response is expected. So after generating and sending the preemptive snapshot, we immediately move on to adding the new replica to the raft group. From the logs, it looks like raft then immediately generates a snapshot to send to the new replica.

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -218,7 +218,11 @@ type Replica struct {
 		// TODO(tschottdorf): evaluate whether this should be a list/slice.
 		pendingCmds       map[storagebase.CmdIDKey]*pendingCmd
 		internalRaftGroup *raft.RawNode
-		replicaID         roachpb.ReplicaID
+		// The ID of the replica within the Raft group. May be 0 if the replica has
+		// been created from a preemptive snapshot (i.e. before being added to the
+		// Raft group). The replica ID will be non-zero whenever the replica is
+		// part of a Raft group.
+		replicaID roachpb.ReplicaID
 		// Most recent timestamps for keys / key ranges.
 		tsCache *timestampCache
 		// proposeRaftCommandFn can be set to mock out the propose operation.
@@ -246,6 +250,11 @@ func (r *Replica) withRaftGroupLocked(f func(r *raft.RawNode) error) error {
 	if r.mu.destroyed != nil {
 		// Silently ignore all operations on destroyed replicas. We can't return an
 		// error here as all errors returned from this method are considered fatal.
+		return nil
+	}
+	if r.mu.replicaID == 0 {
+		// The replica's raft group has not yet been configured (i.e. the replica
+		// was created from a preemptive snapshot).
 		return nil
 	}
 
@@ -351,8 +360,10 @@ func (r *Replica) newReplicaInner(desc *roachpb.RangeDescriptor, clock *hlc.Cloc
 	if replicaID == 0 {
 		_, repDesc := desc.FindReplica(r.store.StoreID())
 		if repDesc == nil {
-			return errors.Errorf("cannot recreate replica that is not a member of its range (StoreID %s not found in %s)",
-				r.store.StoreID(), desc)
+			// This is intentionally not an error and is the code path exercised
+			// during preemptive snapshots. The replica ID will be sent when the
+			// actual raft replica change occurs.
+			return nil
 		}
 		replicaID = repDesc.ReplicaID
 	}

--- a/storage/replica_raftstorage.go
+++ b/storage/replica_raftstorage.go
@@ -534,10 +534,13 @@ func (r *Replica) applySnapshot(snap raftpb.Snapshot) (uint64, error) {
 	replicaID := r.mu.replicaID
 	r.mu.Unlock()
 
-	log.Infof("replica %d received snapshot for range %d at index %d. encoded size=%d, %d KV pairs, %d log entries",
-		replicaID, desc.RangeID, snap.Metadata.Index, len(snap.Data), len(snapData.KV), len(snapData.LogEntries))
+	log.Infof("replica %d received snapshot for range %d at index %d. "+
+		"encoded size=%d, %d KV pairs, %d log entries",
+		replicaID, desc.RangeID, snap.Metadata.Index,
+		len(snap.Data), len(snapData.KV), len(snapData.LogEntries))
 	defer func(start time.Time) {
-		log.Infof("replica %d applied snapshot for range %d in %s", replicaID, desc.RangeID, timeutil.Since(start))
+		log.Infof("replica %d applied snapshot for range %d in %s",
+			replicaID, desc.RangeID, timeutil.Since(start))
 	}(timeutil.Now())
 
 	// Delete everything in the range and recreate it from the snapshot.
@@ -591,6 +594,15 @@ func (r *Replica) applySnapshot(snap raftpb.Snapshot) (uint64, error) {
 	if s.RaftAppliedIndex != snap.Metadata.Index {
 		log.Fatalf("%d: snapshot resulted in appliedIndex=%d, metadataIndex=%d",
 			s.Desc.RangeID, s.RaftAppliedIndex, snap.Metadata.Index)
+	}
+
+	if replicaID == 0 {
+		// The replica is not part of the Raft group so we need to write the Raft
+		// hard state for the replica in order for the Raft state machine to start
+		// correctly.
+		if err := updateHardState(batch, s); err != nil {
+			return 0, err
+		}
 	}
 
 	if err := batch.Commit(); err != nil {

--- a/storage/replica_state.go
+++ b/storage/replica_state.go
@@ -321,6 +321,35 @@ func setHardState(
 		hlc.ZeroTimestamp, nil, &st)
 }
 
+func updateHardState(eng engine.ReadWriter, s storagebase.ReplicaState) error {
+	// Load a potentially existing HardState as we may need to preserve
+	// information about cast votes. For example, during a Split for which
+	// another node's new right-hand side has contacted us before our left-hand
+	// side called in here to create the group.
+	rangeID := s.Desc.RangeID
+	oldHS, err := loadHardState(eng, rangeID)
+	if err != nil {
+		return err
+	}
+
+	newHS := raftpb.HardState{
+		Term:   s.TruncatedState.Term,
+		Commit: s.RaftAppliedIndex,
+	}
+
+	if !raft.IsEmptyHardState(oldHS) {
+		if oldHS.Commit > newHS.Commit {
+			newHS.Commit = oldHS.Commit
+		}
+		if oldHS.Term > newHS.Term {
+			newHS.Term = oldHS.Term
+		}
+		newHS.Vote = oldHS.Vote
+	}
+
+	return setHardState(eng, rangeID, newHS)
+}
+
 // writeInitialState bootstraps a new Raft group (i.e. it is called when we
 // bootstrap a Range, or when setting up the right hand side of a split).
 // Its main task is to persist a consistent Raft (and associated Replica) state
@@ -349,31 +378,7 @@ func writeInitialState(
 		return enginepb.MVCCStats{}, err
 	}
 
-	// Load a potentially existing HardState as we may need to preserve
-	// information about cast votes. For example, during a Split for which
-	// another node's new right-hand side has contacted us before our left-hand
-	// side called in here to create the group.
-	oldHS, err := loadHardState(eng, rangeID)
-	if err != nil {
-		return enginepb.MVCCStats{}, err
-	}
-
-	newHS := raftpb.HardState{
-		Term:   s.TruncatedState.Term,
-		Commit: s.TruncatedState.Index,
-	}
-
-	if !raft.IsEmptyHardState(oldHS) {
-		if oldHS.Commit > newHS.Commit {
-			newHS.Commit = oldHS.Commit
-		}
-		if oldHS.Term > newHS.Term {
-			newHS.Term = oldHS.Term
-		}
-		newHS.Vote = oldHS.Vote
-	}
-
-	if err := setHardState(eng, rangeID, newHS); err != nil {
+	if err := updateHardState(eng, s); err != nil {
 		return enginepb.MVCCStats{}, err
 	}
 

--- a/storage/store.go
+++ b/storage/store.go
@@ -2095,6 +2095,26 @@ func (s *Store) handleRaftMessage(req *RaftMessageRequest) error {
 		return err
 	}
 
+	if req.ToReplica.ReplicaID == 0 {
+		if req.Message.Type == raftpb.MsgSnap {
+			// Allow snapshots to be applied to replicas before they are members of
+			// the raft group (i.e. replicas with an ID of 0). This is the only
+			// operation that can be performed on a replica before it is part of the
+			// raft group.
+			_, err := r.applySnapshot(req.Message.Snapshot)
+			return err
+		}
+		// We disallow non-snapshot messages to replica ID 0. Note that
+		// getOrCreateReplicaLocked disallows moving the replica ID backward, so
+		// the only way we can get here is if the replica did not previously exist.
+		if log.V(1) {
+			log.Infof("refusing incoming Raft message %s for group %d from %+v to %+v",
+				req.Message.Type, req.RangeID, req.FromReplica, req.ToReplica)
+		}
+		return errors.Errorf("cannot recreate replica that is not a member of its range (StoreID %s not found in range %d)",
+			r.store.StoreID(), req.RangeID)
+	}
+
 	if err := r.withRaftGroup(func(raftGroup *raft.RawNode) error {
 		return raftGroup.Step(req.Message)
 	}); err != nil {

